packages/rollup-contracts/contracts/ovm/PartialStateManager.sol

```diff
@@ -0,0 +1,329 @@
+        updatedContractsCounter += 1;
+
+        // Increment the nonce
+        ovmContractNonces[_ovmContractAddress] += 1;
+    }
+
+
+    /*****************
+    * Contract Codes *
+    *****************/
+    // This is used when CALLing a contract
+
+    /**
+     * @notice Attaches some code contract to the desired OVM contract. This allows the Execution Manager
+     *         to later on get the code contract address to perform calls for this OVM contract.
+     * @param _ovmContractAddress The address of the OVM contract we'd like to associate with some code.
+     * @param _codeContractAddress The address of the code contract that's been deployed.
+     */
+    function associateCodeContract(address _ovmContractAddress, address _codeContractAddress) public {
+        require(msg.sender == address(executionManager));
+
+        ovmCodeContracts[_ovmContractAddress] = _codeContractAddress;
+    }
+
+    /**
+     * @notice Lookup the code contract for some OVM contract, allowing CALL opcodes to be performed.
+     * @param _ovmContractAddress The address of the OVM contract.
+     * @return The associated code contract address.
+     */
+    function getCodeContractAddress(address _ovmContractAddress) public returns(address) {
+        ensureVerifiedContract(_ovmContractAddress);
+
+        return ovmCodeContracts[_ovmContractAddress];
+    }
+
+    /**
+     * @notice Get the bytecode at some contract address. NOTE: This is code taken from Solidity docs here:
+     *         https://solidity.readthedocs.io/en/v0.5.0/assembly.html#example
+     * @param _codeContractAddress The address of the code contract.
+     * @return The bytecode at this address.
+     */
+    function getCodeContractBytecode(address _codeContractAddress) public view returns (bytes memory codeContractBytecode) {
+        assembly {
+            // retrieve the size of the code
+            let size := extcodesize(_codeContractAddress)
+            // allocate output byte array - this could also be done without assembly
+            // by using codeContractBytecode = new bytes(size)
+            codeContractBytecode := mload(0x40)
+            // new "memory end" including padding
+            mstore(0x40, add(codeContractBytecode, and(add(add(size, 0x20), 0x1f), not(0x1f))))
+            // store length in memory
+            mstore(codeContractBytecode, size)
+            // actually retrieve the code, this needs assembly
+            extcodecopy(_codeContractAddress, add(codeContractBytecode, 0x20), 0, size)
+        }
+    }
+
+    /**
+     * @notice Get the hash of the deployed bytecode of some code contract.
+     * @param _codeContractAddress The address of the code contract.
+     * @return The hash of the bytecode at this address.
+     */
+    function getCodeContractHash(address _codeContractAddress) public view returns (bytes32 _codeContractHash) {
+        // TODO: Look up cached hash values eventually to avoid having to load all of this bytecode
```

Comment:
yep, this should be authenticated against the state anyway.  And ofc. also note the `EXTCODEHASH` opcode can be used.